demos/bbm_auxiliary/demo_bbm_aux.py.rst

```diff
@@ -0,0 +1,224 @@
+The numerical scheme in this demo introduces
+the :math:`H^1`-Riesz representative of the Fréchet derivative of the
+Hamiltonian :math:`\frac{\delta I_3}{\delta u}` 
+as the auxiliary variable :math:`\tilde{w}_H`.
+
+Standard Gauss-Legendre and continuous Petrov-Galerkin (cPG) methods conserve
+the first two invariants exactly (up to roundoff and solver tolerances).  They
+do quite well, but are inexact for the cubic one.  Here, we consider the
+reformulation in Andrews and Farrell, "Enforcing conservation laws and dissipation
+inequalities numerically via auxiliary variables" (`arXiv:2407.11904 <https://arxiv.org/abs/2407.11904>`_, to appear
+in SIAM J. Scientific Computing) that preserves the third invariant at
+the expense of the second.  This method has an auxiliary variable in the system
+and requires a continuously differentiable spatial discretization (1D Hermite
+elements in this case).  The time discretization puts the main unknown in a
+continuous space and the auxiliary variable in a discontinuous one.  See
+equation (7.17) of Boris Andrews' thesis for the particular formulation.
+
+
+Firedrake, Irksome, and other imports::
+
+  from firedrake import (Constant, Function, FunctionSpace,
+      PeriodicIntervalMesh, SpatialCoordinate, TestFunction, TrialFunction,
+      assemble, derivative, dx, errornorm, exp, grad, inner,
+      interpolate, norm, plot, project, replace, solve, split
+  )
+
+  from irksome import Dt, GalerkinTimeStepper, TimeQuadratureLabel
+
+  import matplotlib.pyplot as plt
+  import numpy
+
+Next, we define the domain and the exact solution ::
+
+  N = 8000
+  L = 100
+  h = L / N
+  msh = PeriodicIntervalMesh(N, L)
+  x, = SpatialCoordinate(msh)
+
+  t = Constant(0)
+  dt = Constant(10*h)
+
+  c = Constant(0.5)
+  center = Constant(40.0)
+  delta = -c * center
+
+  def sech(x):
+      return 2 / (exp(x) + exp(-x))
+  
+  uexact = 3 * c**2 / (1-c**2) \
+      * sech(0.5 * (c * x - c * t / (1 - c ** 2) + delta))**2
+
+This sets up the function space for the unknown :math:`u` and
+auxiliary variable :math:`\tilde{w}_H`::
+
+  space_deg = 3
+  time_deg = 1
+
+  V = FunctionSpace(msh, "Hermite", space_deg)
+  Z = V * V
+
+We next define the BBM invariants. Again, the discrete formulation preserves 
+:math:`I_1` and :math:`I_3` up to solver tolerances and roundoff errors, 
+but :math:`I_2` is preserved up to a bounded oscillation ::
+
+  def h1inner(u, v):
+      return inner(u, v) + inner(grad(u), grad(v))
+
+  def I1(u):
+      return u * dx
+
+  def I2(u):
+      return h1inner(u, u) * dx
+
+  def I3(u):
+      return (u**2 / 2 + u**3 / 6) * dx
+
+We project the initial condition on :math:`u`, but we also need a consistent
+initial condition for the auxiliary variable.  We need to find :math:`\tilde{w}_H \in V` such that
+
+.. math::
+
+   (\tilde{w}_H, v)_{H^1} = \langle \frac{\delta I_3}{\delta u}, v \rangle \text{ for all } v \in V
+
+::
+
+  uwHtilde = Function(Z)
+  uinit, wHinit = uwHtilde.subfunctions
```

Comment:
We redefine `uwHtilde.subfunctions` on line 151.  Why not just use `u, wH` here?

Reply:
Because that will break `replace(dHdu, {uinit: u}`.